fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -458,4 +483,207 @@ public static long getFrom(@Nonnull LucenePartitionInfoProto.LucenePartitionInfo
+        Collection<RecordType> recordTypes = state.store.getRecordMetaData().recordTypesForIndex(state.index);
+        if (recordTypes.stream().map(RecordType::isSynthetic).distinct().count() > 1) {
+            // don't support mix of synthetic/regular
+            throw new RecordCoreException("mix of synthetic and non-synthetic record types in index is not supported");
+        }
+
+        final CompletableFuture<? extends List<? extends FDBIndexableRecord<Message>>> fetchedRecordsFuture;
+        if (recordTypes.iterator().next().isSynthetic()) {
+            fetchedRecordsFuture = cursor.mapPipelined(indexEntry -> state.store.loadSyntheticRecord(indexEntry.getPrimaryKey()),
+                    state.store.getPipelineSize(PipelineOperation.INDEX_TO_RECORD)).asList();
+        } else {
+            fetchedRecordsFuture = state.store.fetchIndexRecords(cursor, IndexOrphanBehavior.SKIP).map(FDBIndexedRecord::getStoredRecord).asList();
+        }
+
+        return fetchedRecordsFuture.thenCompose(records -> {
+            if (records.size() > 0) {
+                // the newest record is the one we intend to leave in the current partition; we need it in order to set this partition's new
+                // `from` value.
+                final long newBoundaryTimestamp = getPartitioningTimestampValue(records.get(records.size() - 1));
+
+                // remove the (n + 1)th record from the records to be moved
+                records.remove(records.size() - 1);
+
+                LuceneIndexMaintainer indexMaintainer = (LuceneIndexMaintainer)state.store.getIndexMaintainer(state.index);
+                // shortcut delete docs from current partition
+                // (we do this, instead of calling LuceneIndexMaintainer.update() in order to avoid a chicken-and-egg
+                // situation with the partition metadata keys.
+                records.forEach(r -> {
+                    try {
+                        indexMaintainer.deleteDocument(groupingKey, partitionInfo.getId(), r.getPrimaryKey());
+                    } catch (IOException e) {
+                        throw new RecordCoreException(e);
+                    }
+                });
+
+                // update current partition's meta
+                state.context.ensureActive().clear(partitionMetadataKeyFromTimestamp(groupingKey, getFrom(partitionInfo)));
+                LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partitionInfo.toBuilder()
+                        .setCount(partitionInfo.getCount() - records.size())
+                        .setFrom(ByteString.copyFrom(Tuple.from(newBoundaryTimestamp).pack()));
+                savePartitionMetadata(groupingKey, builder);
+
+                // value of the "destination" partition's `from` timestamp
+                final long overflowPartitionFromTimestamp = getPartitioningTimestampValue(records.get(0));
```

Comment:
What happens if there are ties? In particular, if some after the end of the cursor have the same timestamp? If this is a failure of the caller to obey the contract of the timestamp data can we at least detect that?

Reply:
Yes, @wshaib-apple and I had discussed this, and it doesn't support that. it probably makes sense to at least throw an error at this point, and that should be fairly easy and worth bringing in now.
It would also be pretty easy to support any situation where you have fewer than `LuceneRecordContextProperties.LUCENE_REPARTITION_DOCUMENT_COUNT` documents with the same timestamp. 
To support the broader case my two thoughts (right now) would be to:
1. Add the primary key to the `from` so that you're guaranteed to have unique entries.
2. Allow the from of one partition to be the same as the to in the other and do something else to ensure that each document is only in one partition.

If you have more than a very small ratio of `LuceneIndexOptions.INDEX_PARTITION_HIGH_WATERMARK` to documents with the same timestamp, the whole feature kind of falls over, but I don't think we need to require the timestamp to be unique.